Workflow Operation "CutMarksToSmil"

Arnei opened the pull request:
This operation parses a JSON containing cut marks into a SMIL that can be used by the VideoEditorWorkflowOperation. It does this by attributing the given times to the tracks in the given presentation and presenter flavors. This is useful in case you have cutmarks that need to be applied to files that are created during a workflow.

Use-case example: Sending raw recording data from a web-conferencing application to Opencast, where Opencast should first concatenate the data and then cut the concatenated video file by user-defined cut marks.

An example workflow of how this can be used can be found in the [bbb-upload.xml in this repo](https://github.com/elan-ev/opencast-bigbluebutton-integration/tree/master/post-archive).

### Your pull request should…

* [x] have a concise title
* [x] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [x] be against the correct branch (features can only go into develop)
* [x] include migration scripts and documentation, if appropriate
* [x] pass automated testing
* [x] have a clean commit history
* [x] have proper commit messages (title and body) for all commits
* [x] have appropriate tags applied

ebbertd commented on docs/guides/admin/docs/workflowoperationhandlers/cut-marks-to-smil-woh.md:
```diff
@@ -0,0 +1,66 @@
+|------------------         |-----------            |---------------------------------------------------------------|
+|source-presenter-flavor    |`presenter/prepared`   |The flavor of the presenter video track. Must contain exactly one file.                   |
+|source-presentation-flavor |`presentation/prepared`|The flavor of the presentation video track. Must contain exactly one file.                  |
+|source-smil-flavor         |`smil/times`           |The flavor of the JSON. Must contain exactly one file.|
```
Naming this configuration key `source-smil-flavor` is confusing. Based on this name I would expect a smil file, which is xml. However, you expect a JSON with the cut marks from BBB. Therefore I suggest to name it `source-json-flavor` and to amend the description with the information that these are the cut marks from BBB.

Arnei replied:
Yup, that makes sense.
I would like to keep BBB out of the description, because I wouldn't want people to think that this WHO can only be used with BBB.

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/cut-marks-to-smil-woh.md:
```diff
@@ -0,0 +1,66 @@
+|configuration keys         |example                |description                                                    |
+|------------------         |-----------            |---------------------------------------------------------------|
+|source-presenter-flavor    |`presenter/prepared`   |The flavor of the presenter video track. Must contain exactly one file.                   |
+|source-presentation-flavor |`presentation/prepared`|The flavor of the presentation video track. Must contain exactly one file.                  |
```
Can we not hard-code presentation and presenter but just use `source-flavors` where you can define a list of flavors in the form `a/x,b/x` or a wildcard like `*/source`? That should be negligently harder to do and makes this far more flexible.

Arnei replied:
I changed it to support a list of flavors (`a/x,b/x`). Does not support wildcards, because I could not figure out how to derive a list of flavors from them. Which I would like to do to check for the amount of tracks in each flavor, because the WHO does not know how it should handle multiple tracks per flavor.

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/cut-marks-to-smil-woh.md:
```diff
@@ -0,0 +1,66 @@
+|target-smil-flavor         |`smil/cutmarks`        |The flavor of the resulting SMIL.|
+
+## JSON Format
+*begin* marks the start of a segment, *duration* its duration. Times are in milliseconds.
```
```suggestion
The JSON structure specifies all segments which should be kept after cutting.
The property `begin` marks the start of a segmen while `duration` its duration.
The values are specified in milliseconds.
```

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/cut-marks-to-smil-woh.md:
```diff
@@ -0,0 +1,66 @@
+            <configuration key="source-json-flavor">smil/times</configuration>
+            <configuration key="target-smil-flavor">smil/cutmarks</configuration>
+          </configurations>
+        </operation>
```
- Keep the example as simple as possible
- Directly generate `smil/cutting`
- Use an XML code block

```suggestion
```xml
<operation
    id="cut-marks-to-smil"
    description="Process ingested cutmarks by applying them to current tracks">
  <configurations>
    <configuration key="source-presenter-flavor">presenter/prepared</configuration>
    <configuration key="source-presentation-flavor">presentation/prepared</configuration>
    <configuration key="source-json-flavor">smil/times</configuration>
    <configuration key="target-smil-flavor">smil/cutting</configuration>
  </configurations>
</operation>
```
```

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/cut-marks-to-smil-woh.md:
```diff
@@ -0,0 +1,66 @@
+            <configuration key="target-smil-flavor">smil/cutting</configuration>
+            <configuration key="target-flavor-subtype">trimmed</configuration>
+          </configurations>
+        </operation>
```
Not sure if we need half a workflow here
```suggestion
```

lkiesow commented on etc/workflows/bbb-upload.xml:
```diff
@@ -0,0 +1,291 @@
+  </tags>
+  <displayOrder>1000</displayOrder>
+  <description>
+    Knock! Knock! Who's there? Candice. Candice who? Candice door open, or am I stuck out here?
```
This probably belongs not in here but in the BBB integration documentation at https://github.com/elan-ev/opencast-bigbluebutton-integration/tree/master/post-archive

Though with no or a sensible description ;p

Arnei replied:
Agreed, the workflow was not intended as part of the merge request, but for potential testers to have a workflow at the ready. I'll remove it then and replace it with a link to the repo you mentioned.

...and add an actual description.

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,326 @@
+        try {
+          workspace.delete(elem.getURI());
+        } catch (Exception e) {
+          logger.warn("Unable to delete element {}: {}", elem, e);
```
```suggestion
          logger.warn("Unable to delete element {}", elem, e);
```

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,326 @@
+    // Is there a catalog?
+    Catalog[] catalogs = mediaPackage.getCatalogs(jsonFlavor);
+    if (catalogs.length != 1) {
+      logger.warn("Number of catalogs in the source flavor does not equal one. Skipping...");
+      return skip(mediaPackage);
```
I tend to think we should rather fail here.
Otherwise, we might accidentally publish uncut material without anyone notiving

Arnei replied:
The choice to skip instead of fail if no (or too many) catalogs are provided was made to be consistent with the behaviour of the VideoEditor WHO, which also does not fail if no cataloge is provided. Therefore I'd rather not fail here, although maybe it would be better to fail if there >1 cataloges.

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,326 @@
+
+    // Check parsing results
+    for (Times entry : cutmarks) {
+      logger.info("Entry begin {}, Entry duration {}", entry.begin, entry.duration);
```
```suggestion
      logger.debug("Entry begin {}, Entry duration {}", entry.begin, entry.duration);
```

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,326 @@
+      }
+
+      Smil smil = smilResponse.getSmil();
+      logger.info("Done Adding tracks");
```
```suggestion
      logger.info("Done adding tracks");
```

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,326 @@
+      Smil smil = smilResponse.getSmil();
+      logger.info("Done Adding tracks");
+      InputStream is = null;
+      try {
```
Maybe use Java's [try-with-resources statement](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,326 @@
+    final WorkflowOperationResult result = createResult(mediaPackage, WorkflowOperationResult.Action.CONTINUE);
+    logger.debug("Partial import operation completed");
+    return result;
+  }
```
Why re-define _skip_ as _continue_? That seems wrong.
I dont' actually see the purpose of this method at all.

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,326 @@
+    } catch (NotFoundException e) {
+      throw new WorkflowOperationException(
+              "Error finding the media file in the workspace", e);
+    } catch (IOException e) {
```
```suggestion
    } catch (NotFoundException | IOException e) {
```

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    String filePath = mediaFile.getAbsolutePath();
+    return filePath;
```
```suggestion
    return mediaFile.getAbsolutePath();
```

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,326 @@
+ */
+package org.opencastproject.workflow.handler.composer;
+
+import static com.entwinemedia.fn.Stream.$;
```
We passed a rule that we won't accept new code using the Entwine Functional Library unless it really can't be avoided. For this pull request, you can easily use other means to archive the same goal.
Please avoid it.

Arnei replied:
Understood, thanks for the heads up.

lkiesow commented:
The build failure is caused by misformatting:

> [ERROR] /home/travis/build/opencast/opencast/modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:268: Line matches the illegal pattern 'Line has trailing spaces.'. [Regexp]

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/cut-marks-to-smil-woh.md:
```diff
@@ -0,0 +1,44 @@
+CutMarksToSmilWorkflowOperationHandler
```
This is missing from the [list of workflow operation handlers](https://docs.opencast.org/develop/admin/#workflowoperationhandlers/#default-workflow-operations)

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/cut-marks-to-smil-woh.md:
```diff
@@ -0,0 +1,44 @@
+The property `begin` marks the start of a segment while `duration` its duration.
+The values are specified in milliseconds.
+
+    [
+      {
+        "begin": 1672,
+        "duration": 7199
+      }
+    ]
```
```suggestion
```json
[
  {
    "begin": 1672,
    "duration": 7199
  }
]
```
```

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,319 @@
+
+    // Is there a catalog?
+    Catalog[] catalogs = mediaPackage.getCatalogs(jsonFlavor);
+    if (catalogs.length != 1) {
+      logger.warn("Number of catalogs in the source flavor does not equal one. Skipping...");
```
Skipping this if there is none may be fine but skipping it if there are multiple catalogs is weird and unexpected. I would expect that they are both converted or that there is an error.

Arnei replied:
Right, that makes sense. I'll have it throw an error in that case for now, since processing multiple catalogs seems non-trivial (e.g. what if there are overlapping cutmarks).

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,319 @@
+    try (BufferedReader bufferedReader = new BufferedReader(new FileReader(getMediaPackageElementPath(jsonWithTimes)))) {
+      cutmarks = gson.fromJson(bufferedReader, Times[].class);
+    } catch (Exception e) {
+      throw new WorkflowOperationException("Could not read JSON: " + e);
```
Use the `cause` argument of exception constructors:
```suggestion
      throw new WorkflowOperationException("Could not read JSON", e);
```

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,319 @@
+    for (Times entry : cutmarks) {
+      logger.info("Entry begin {}, Entry duration {}", entry.begin, entry.duration);
+      if (entry.begin < 0 || entry.duration < 0) {
+        throw new WorkflowOperationException("Times cannot be negative!");
```
```suggestion
        throw new WorkflowOperationException("Times may not be negative.");
```

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,319 @@
+    }
+
+    // Get video tracks
+    logger.info("Get Tracks from Mediapackage");
```
```suggestion
    logger.info("Get tracks from media package");
```

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,319 @@
+      return result;
+    if (videos.size() > 1) {
+      for (Track track : videos) {
+        logger.info("Track {} in flavor {}", track, flavor);
```
A flavor is no collection. It's an attribute:
```suggestion
        logger.info("Track {} with flavor {}", track, flavor);
```

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,319 @@
+      for (Track track : videos) {
+        logger.info("Track {} in flavor {}", track, flavor);
+      }
+      throw new WorkflowOperationException("The number of videos in each flavor cannot be more than one.");
```
Is there a reason why there can be only one video with a given flavor in this operation?

Arnei replied:
Iirc the reason for that was that the behaviour for multiple tracks in the same flavor is undefined in the editor. So this basically acts as an early error check.

lkiesow replied:
That sounds like a bug in the editor. For some operations, you actually need multiple tracks with identical flavors. Letting it break here and kind of building more bugs doesn't sound right ;-)

Arnei replied:
That does make sense. I just did some testing on the matter and having multiple videos in one flavor seems to work just fine. Applied your suggest change with commit 6e9702c .

lkiesow commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,319 @@
+    try {
+      SmilResponse smilResponse = smilService.createNewSmil(mediaPackage);
+
+      logger.info("Start Adding tracks");
```
```suggestion
      logger.info("Start adding tracks");
```

gregorydlogan commented:
@Arnei You're hoping to get this into the codebase correct?  It's ready for review, or are you waiting on other things?

Arnei commented:
@gregorydlogan Yes, I do hope to get this into the codebase and it is ready for review.

lkiesow reviewed (commented):
As discussed earlier, let's put this into a separate module since it does not really have to do anything with the composer. That's something I can prepare. This will hopefully also allow us to tidy up a bit.

Arnei committed:
WorkflowOperationHandler "CutMarksToSmil"

This operation parses a JSON containing cut marks into a SMIL that can
be used by the VideoEditorWorkflowOperation. It does this by attributing
the given times to the tracks in the given presentation and presenter
flavors. This is useful in case you have cutmarks that need to be
applied to files that are created during a workflow.

Use-case example: Sending raw recording data from a web-conferencing
application to Opencast, where Opencast should first concatenate the
data and then cut the concatenated video file by user-defined cut marks.

Also adds a workflow showcasing the functionality.

Co-authored-by: Lars Kiesow <lkiesow@uos.de>

lkiesow committed:
Improve Cut Marks to Smil Operation

This patch fixes and improves the cut-marks to Smil operation. It's
mainly minor code fixes and a few dependency fixes. This should not have
changed the overall functionality.

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/cut-marks-to-smil-woh.md:
```diff
@@ -0,0 +1,57 @@
+    exception-handler-workflow="partial-error">
+  <configurations>
+    <configuration key="source-media-flavors">presenter/prepared,presentation/prepared</configuration>
+    <configuration key="source-json-flavor">json/times</configuration>
```
I'm wondering if something like `cut-marks/json` wouldn't be a better default name.

Arnei replied:
"cut-marks" instead of "times" seems more specific and therefore clearer to understand, so +1 for that. As for the order, I put the file type first to be consistent with naming of SMIL catalogue flavors (e.g. `smil/source+partial`), but I got no hard opinion on this.

lkiesow replied:
I decided to try `cut-marks/json`. The `smil/*` is a good argument, but I don't think that smil is really meant as a file type here since it's something specific and not a general purpose file type like JSON. That would make the main type otherwise very unspecific and a wildcard like `json/*` could catch all kinds of things.

 At the same time, we also have other favors with file format as sub-type:
 
- `security/xacml(+…)`
- `captions/google-speech-json`
- `captions/ibm-watson-json`

In the end, it's just the documentation, and you can pick whatever you want anyway. Take a look, and we can still change it if you want to. I'm not that fixed on this. I will push the update shortly.

lkiesow commented on modules/smil-workflowoperation/src/main/java/org/opencastproject/workflow/handler/smil/CutMarksToSmilWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,303 @@
+    trackSelector.addFlavor(flavor);
+    Collection<Track> tracks = trackSelector.select(mediaPackage, false);
+
+    // Get only videos
+    ArrayList<Track> videos = new ArrayList<>();
+    for (Track video : tracks) {
+      if (video.hasVideo()) {
+        videos.add((video));
+      } else {
+        logger.debug("Skipping track {} since it does not seem to have a vide track.", video);
+      }
+    }
```
I'm wondering if there is a specific reason to require video streams to be present. This would mean that we cannot cut audio and also that we need to make sure that we run `inspect` on the media package before we run this operation.

Arnei replied:
I honestly do not remember why only video tracks were allowed (should've commented on that, arg). I'll try and see if I can't figure it out.

Arnei replied:
Whatever the reason was that I originally implemented this, it seems to have vanished.  So I suppose this code snippet could be removed.

lkiesow replied:
Removed the filter. Now it's just the track selector.

lkiesow commented on modules/smil-workflowoperation/src/main/resources/OSGI-INF/operations/cut-marks-to-smil.xml:
```diff
@@ -0,0 +1,17 @@
+               immediate="true">
+  <implementation class="org.opencastproject.workflow.handler.smil.CutMarksToSmilWorkflowOperationHandler"/>
+  <property name="service.description" value="Cut Marks To Smil Operation Handler"/>
+  <property name="workflow.operation" value="cut-marks-to-smil"/>
```
If anyone else is wondering: It would be nice to not specify this in XML but unfortunately, some methods are contained in the parent class which means we would probably need to change all the operation handlers at once.

lkiesow reviewed (commented):
As discussed, I've moved this to a separate module and fixed a few additional minor issues (see second commit). There are two remaining open questions I have for you, @Arnei. Apart from that, I'm happy with merging this.

For testing this, here are some test scripts (make sure you have a vide.mp4 in the script folder):
https://data.lkiesow.io/opencast/cut-marks-test-data/

lkiesow committed:
Cut Marks to Smil Documentation

This patch adds some additional documentation about the new workflow
operation handler.

Arnei commented:
> As discussed, I've moved this to a separate module and fixed a few additional minor issues (see second commit). 

Looks good to me.

lkiesow committed:
Apply Review Suggestions

This patch contains some slight improvements to the `cut-marks-to-smil`
operation handler.

lkiesow commented:
@Arnei, can you check a final time if you are happy with the current state? I will merge this if you are.

Arnei commented:
> @Arnei, can you check a final time if you are happy with the current state? I will merge this if you are.

Looks good to me and seems to be in working order.

lkiesow commented:
Yay, complains now about the new checkstyle rules… fixing. Then merging…
